File: Lv.2/오픈채팅방.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
#include <functional>
#include <sstream>
#include <map>
using namespace std;

vector<string> solution(vector<string> record) {
    vector<string> answer;

    vector<pair<string,string>> cmdId; // 명령어와 id값을 pair로 묶어서 vector에 저장
    map<string,string> idName; // id값과 그에 해당하는 name를 관리할 map

    // record를 바탕으로, cmdId와 idName 세팅하기
    for(int i=0;i<record.size();i++){
        string temp = record[i];
        stringstream ss(temp);
        string cmd; // 수행동작
        string id; // 회원의 id
        string name; // 회원의 이름
        ss>>cmd>>id>>name;
        
        // 입장했을 경우
        if(cmd=="Enter"){
            
            cmdId.push_back(make_pair(cmd,id)); // 순서를 위해 vector에 저장
            idName[id]=name; // 현재 id에 해당하는 name값 업데이트
            
        }
        // 나갔을 경우
        else if(cmd=="Leave"){ 
            cmdId.push_back(make_pair(cmd,id)); // 순서를 위해 vector에 저장
        }
        
        // 이름 변경했을 경우
        else{ // Change
            idName[id]=name;  // 현재 id에 해당하는 name값 업데이트
        }

    }
    
    // 값 출력
    for(auto k : cmdId){
        if(k.first == "Enter"){
            answer.push_back((idName[k.second])+"님이 들어왔습니다.");
        }
        else if(k.first == "Leave"){
            answer.push_back((idName[k.second])+"님이 나갔습니다.");
        }
    }


    return answer;
}

// enter 1234  1234 prodo
// enter 4567  4567 ryan
// leave 1234  
// enter 1234
// change 4567

File: Java/LV2/오픈채팅방.java
import java.util.*;

class Pair{
    String first;
    String second;
    
    Pair(String x, String y){
        first=x;
        second=y;
    }
}

class Solution {
    public String[] solution(String[] record) {
        String[] answer;
        
        ArrayList<Pair> cmdId = new ArrayList<Pair>();
        HashMap<String,String> idName = new HashMap<>();
        
        for(int i=0;i<record.length;i++){
            String s = record[i];
            String[] temp = s.split(" ");
            
            String cmd = temp[0];
            String id = temp[1];
            String name = "";
            
            if(!cmd.equals("Leave")){
                name = temp[2];
            }
            
            if(cmd.equals("Enter")){
                
                idName.put(id,name);
                cmdId.add(new Pair(cmd,id));
            }
            else if(cmd.equals("Leave")){
                cmdId.add(new Pair(cmd,id));
            }
            else{
                
                idName.put(id,name);
            }
            
            
            
            
        }
        answer = new String[cmdId.size()];
            for(int i=0;i<cmdId.size();i++){
                if(cmdId.get(i).first.equals("Enter")){
                    String tempName = idName.get(cmdId.get(i).second);
                    answer[i]=tempName+"님이 들어왔습니다.";  
                }
                else if(cmdId.get(i).first.equals("Leave")){
                    String tempName = idName.get(cmdId.get(i).second);
                    answer[i]=tempName+"님이 나갔습니다.";
                }
                
            }
        return answer;
        
        
    }
}
